packages/next/src/lib/metadata/resolve-opengraph.ts

```diff
@@ -0,0 +1,88 @@
+import type { Metadata, ResolvedMetadata } from './types/metadata-interface'
+import type { OpenGraphType, OpenGraph } from './types/opengraph-types'
+
+function resolveAsArrayOrUndefined<T = any>(value: T): undefined | any[] {
+  if (typeof value === 'undefined' || value === null) {
+    return undefined
+  }
+  if (Array.isArray(value)) {
+    return value
+  }
+  return [value]
+}
+
+function getFieldsByOgType(ogType: OpenGraphType | undefined) {
+  const articleKeys = ['authors', 'tags'] // article, book
+  const songKeys = ['albums', 'musicians'] // music.song, music.album
+  const playlistKeys = ['albums', 'musicians'] // music.playlist
+  const radioStationsKeys = ['creators']
+  const videoKeys = ['actors', 'directors', 'writers', 'tags']
+
+  switch (ogType) {
+    case 'article':
+    case 'book':
+      return articleKeys
+    case 'music.song':
+    case 'music.album':
+      return songKeys
+    case 'music.playlist':
+      return playlistKeys
+    case 'music.radio_station':
+      return radioStationsKeys
+    case 'video.movie':
+    case 'video.episode':
+      return videoKeys
+    default:
+      return [
+        'emails',
+        'phoneNumbers',
+        'faxNumbers',
+        'alternateLocale',
+        'images',
+        'audio',
+        'videos',
+      ]
+  }
+}
+
+export function resolveOpenGraph(
+  openGraph: Metadata['openGraph']
+): ResolvedMetadata['openGraph'] {
+  const url = openGraph
+    ? typeof openGraph.url === 'string'
+      ? new URL(openGraph.url)
```

Comment:
This is a case where `metadataBase` should apply I believe. It is my understanding that urls in og tags need to be fully qualified. The logic maybe should be something like this

```
// this URL does not have a trailing slash but we should normalize it to assume it always ends in a "directory" not a "file" since we are going to extend it
baseURL = new URL("https://foo.example/base")
// normalize to directory
if (!baseURL.pathName.endsWith('/')) {
  baseURL.pathname = baseURL.pathname + '/';
}

url = new URL(urlStringOrURL, baseURL);
```

This should map the following
```
baseURL = new URL("https://foo.example/base")
"bar" => new URL("https://foo.example/base/bar")
"/bar" => new URL("https://foo.example/base/bar")
"./bar" => new URL("https://foo.example/base/bar")
"../bar" => new URL("https://foo.example/bar")
"https://bar.example/bar" => new URL("https://bar.example/bar")
new URL("https://bar.example/bar") => new URL("https://bar.example/bar")
```

Reply:
🙏  Will address the metadataBase in the following PRs! thanks for the details 👍